nemo/collections/nlp/modules/common/megatron/megatron_bert.py

```diff
@@ -49,52 +57,53 @@ class MegatronBertEncoder(BertModule):
     """
 
     def __init__(
-        self,
-        model_name,
-        vocab_file,
-        hidden_size=1024,
-        num_attention_heads=16,
-        num_layers=24,
-        max_seq_length=512,
-        tokenizer_type='BertWordPieceLowerCase',
-        init_method_std=0.02,
-        num_tokentypes=2,
+        self, model_name, config,
     ):
 
         super().__init__()
 
-        if not os.path.exists(vocab_file):
-            raise ValueError(f'Vocab file not found at {vocab_file}')
-
-        megatron_args = {
-            "num_layers": num_layers,
-            "hidden_size": hidden_size,
-            "num_attention_heads": num_attention_heads,
-            "max_position_embeddings": max_seq_length,
-            "tokenizer_type": tokenizer_type,
-            "vocab_file": vocab_file,
-        }
-
-        self.config = BertConfig(
-            num_layers=num_layers,
-            hidden_size=hidden_size,
-            num_attention_heads=num_attention_heads,
-            tokenizer_type=tokenizer_type,
-            vocab_file=vocab_file,
-        )
+        if not os.path.exists(config["vocab_file"]):
+            raise ValueError(f'Vocab file not found at {config["vocab_file"]}')
+
+        config['tokenizer_type'] = 'BertWordPieceLowerCase'
+        set_global_variables(extra_args_provider=None, args_defaults=config, ignore_unknown_args=True)
+        # read arguments back
+        args = get_args()
 
-        initialize_megatron(None, megatron_args, ignore_unknown_args=True)
-        init_method = init_method_normal(init_method_std)
 
-        self.language_model, self._language_model_key = get_language_model(
+        if args.distributed_backend == 'ddp':
+            set_model_parallel_rank(args.rank)
+            set_model_parallel_world_size(args.model_parallel_size)
+        else:
+            _initialize_distributed()
+
+        # Autoresume.
+        _init_autoresume()
+
+        # Random seeds for reproducibility.
+        if args.rank == 0:
+            seed = args.seed
+            if seed is not None and seed > 0:
+                print('> setting random seeds to {} ...'.format(args.seed))
+                random.seed(seed)
+                np.random.seed(seed)
+                torch.manual_seed(seed)
+
+        print(f"MegatronBertEncoder.init: rank={args.rank}, model_parallel_size = {args.model_parallel_size}, dist={dist.is_initialized()}")
+            
+        self.language_model = TransformerLanguageModel(
             attention_mask_func=bert_attention_mask_func,
-            num_tokentypes=num_tokentypes,
+            mlp_activation_func=torch.nn.functional.gelu,
+            init_method=init_method_normal(args.init_method_std),
+            output_layer_init_method=scaled_init_method_normal(args.init_method_std, args.num_layers),
+            num_tokentypes=2,
```

Comment:
So basically here you're reproducing get_language_model() but fixing args along the way?

Reply:
Yes.